Network Observability API documentation updates (#71033)
openshift/openshift-docs: modules/network-observability-flowcollector-api-specifications.adoc

// Automatically generated by 'openshift-apidocs-gen'. Do not edit.

> :_content-type: REFERENCE

Reviewer: is that something to change upstream?

| `apiVersion`
| `string`

> | APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources

Reviewer: Should we use "might" instead of "may" upstream?

Author: Yes

| `object`
| Defines the desired state of the FlowCollector resource.  +
 +

>  *: the mention of "unsupported", or "deprecated" for a feature throughout this document means that this feature is not officially supported by Red Hat. It might have been, for instance, contributed by the community and accepted without a formal agreement for maintenance. The product maintainers might provide some support for these features as a best effort only.

Reviewer: same question

| `privileged`
| `boolean`

> | Privileged mode for the eBPF Agent container. When ignored or set to `false`, the operator sets granular capabilities (BPF, PERFMON, NET_ADMIN, SYS_RESOURCE) to the container. If for some reason these capabilities cannot be set, such as if an old kernel version not knowing CAP_BPF is in use, then you can turn on this mode for more global privileges. Some agent features require the privileged mode, such as packet drops tracking (see `features`).

Reviewer 2: Yes, we should add for SRIOV as well.

> | Privileged mode for the eBPF Agent container. When ignored or set to `false`, the operator sets granular capabilities (BPF, PERFMON, NET_ADMIN, SYS_RESOURCE) to the container. If for some reason these capabilities cannot be set, such as if an old kernel version not knowing CAP_BPF is in use, then you can turn on this mode for more global privileges. Some agent features require the privileged mode, such as packet drops tracking (see `features`).

Reviewer 3: @skrthomas the text cold be something like 
```suggestion
Some agent features require the privileged mode, such as packet drops tracking (see `features`) and SR-IOV support.
```

|===
| Property | Type | Description

| `advanced`
| `object`
| `advanced` allows setting some aspects of the internal configuration of the Loki clients. This section is aimed mostly for debugging and fine-grained performance optimizations.

| `enable`
| `boolean`
| Set `enable` to `true` to store flows in Loki. It is required for the {product-title} Console plugin installation.

| `lokiStack`
| `object`

> | Loki configuration for "LokiStack" mode. This is useful for an easy loki-operator configuration. It is ignored for other modes.

Author: Same for this mode. I think if we mention it, should it be in ` ` and also match the casing that a user would need to use in the YAML? Like `lokiStack`?

Reviewer: no the case here is important it's "LokiStack" for the value of `mode` ; and there's a key named `lokiStack`. I think it's good as it is

| `manual`
| `object`
| Loki configuration for "Manual" mode. This is the most flexible configuration. It is ignored for other modes.

| `microservices`
| `object`
| Loki configuration for "Microservices" mode. Use this option when Loki is installed using the microservices deployment mode (https://grafana.com/docs/loki/latest/fundamentals/architecture/deployment-modes/#microservices-mode). It is ignored for other modes.

| `mode`
| `string`
| `mode` must be set according to the installation mode of Loki: +
 - Use "LokiStack" when Loki is managed using the Loki Operator +
 - Use "Monolithic" when Loki is installed as a monolithic workload +
 - Use "Microservices" when Loki is installed as microservices, but without Loki Operator +

>  - Use "Manual" if none of the options above match your setup +

Author: @jotak I think we should put these `mode` values in ` ` instead of " "

Reviewer: hmm I thought we were using " " when we refer to string yaml values, as opposed to ` ` for keys / paths and perhaps a few other things ...
I need to check if we do that in other places to be consistent

Reviewer: ok, you're right :-) I'll change upstream

| `monolithic`
| `object`

> | Loki configuration for "Monolithic" mode. Use this option when Loki is installed using the monolithic deployment mode (https://grafana.com/docs/loki/latest/fundamentals/architecture/deployment-modes/#monolithic-mode). It is ignored for other modes.

Author: Same question here for `monolithic` vs "Monolithic"

| `writeBatchSize` is the maximum batch size (in bytes) of Loki logs to accumulate before sending.

| `writeBatchWait`
| `string`
| `writeBatchWait` is the maximum time to wait before sending a Loki batch.

| `writeTimeout`
| `string`
| `writeTimeout` is the maximum Loki time connection / request limit. A timeout of zero means no timeout.

|===
== .spec.loki.advanced
Description::
+
--
`advanced` allows setting some aspects of the internal configuration of the Loki clients. This section is aimed mostly for debugging and fine-grained performance optimizations.
--

Type::
  `object`

[cols="1,1,1",options="header"]
|===
| Property | Type | Description

| `staticLabels`
| `object (string)`
| `staticLabels` is a map of common labels to set on each flow in Loki storage.

| `writeMaxBackoff`
| `string`
| `writeMaxBackoff` is the maximum backoff time for Loki client connection between retries.

| `writeMaxRetries`
| `integer`
| `writeMaxRetries` is the maximum number of retries for Loki client connections.

| `writeMinBackoff`
| `string`
| `writeMinBackoff` is the initial backoff time for Loki client connection between retries.

|===
== .spec.loki.lokiStack
Description::
+
--

> Loki configuration for "LokiStack" mode. This is useful for an easy loki-operator configuration. It is ignored for other modes.

Author: Same question here for `lokiStack` vs "LokiStack"

| `logTypes`
| `string`
| `logTypes` defines the desired record types to generate. Possible values are: +
 - `Flows` (default) to export regular network flows +
 - `Conversations` to generate events for started conversations, ended conversations as well as periodic "tick" updates +
 - `EndedConversations` to generate only ended conversations events +
 - `All` to generate both network flows and all conversations events +

| `metrics`
| `object`
| `Metrics` define the processor configuration regarding metrics

| `multiClusterDeployment`
| `boolean`

> | Set `multiClusterDeployment` to `true` to enable multi clusters feature. This adds `clusterName` label to flows data

Reviewer 3: ```suggestion
| Set `multiClusterDeployment` to `true` to enable multi-cluster feature. This adds `clusterName` label to flows data
```

Author: According to IBM style guide:
![image](https://github.com/openshift/openshift-docs/assets/77501157/98ea32ab-619c-416b-aa87-6c1ce3465ed8)
<img width="1036" alt="image" src="https://github.com/openshift/openshift-docs/assets/77501157/a906cff2-c91d-49c0-9ab1-8f659f99f192">

I take this guidance, along with a search of "multi-cluster" and "multicluster" in the IBM style guide, to mean that "multicluster" is the correct way to form this word. In my search, I see "multicluster" with no hyphen.

Reviewer 3: @skrthomas you would know better than I! Good with me :+1:

| `resources`
| `object`
| `resources` are the compute resources required by this container. More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/

|===
== .spec.processor.advanced
Description::
+
--

> `advanced` allows setting some aspects of the internal configuration of the flow processor. This section is aimed mostly for debugging and fine-grained performance optimizations, such as `GOGC` and `GOMAXPROCS` env vars. Users setting its values do it at their own risk.

Reviewer 3: ```suggestion
`advanced` allows setting some aspects of the internal configuration of the flow processor. This section is aimed mostly for debugging and fine-grained performance optimizations, such as `GOGC` and `GOMAXPROCS` env vars. Set these values at your own risk.
```

Reviewer: @skrthomas does this change look good to you? If so I'll take it upstream

Author: Yep, these suggestions look good to me.

 `NetObservLokiError`, which is triggered when flows are being dropped due to Loki errors. +

| `includeList`
| `array (string)`

> | `includeList` is a list of metric names to specify which ones to generate. The names correspond to the names in Prometheus without the prefix. For example, `namespace_egress_packets_total` shows up as `netobserv_namespace_egress_packets_total` in Prometheus. Note that the more metrics you add, the bigger is the impact on Prometheus workload resources. Metrics enabled by default are: `namespace_flows_total`, `node_ingress_bytes_total`, `workload_ingress_bytes_total`, `namespace_drop_packets_total` (when `PacketDrop` feature is enabled), `namespace_rtt_seconds` (when `FlowRTT` feature is enabled), `namespace_dns_latency_seconds` (when `DNSTracking` feature is enabled). More information, with full list of available metrics: https://github.com/netobserv/network-observability-operator/blob/main/docs/Metrics.md

Reviewer: I created a jira: https://issues.redhat.com/browse/NETOBSERV-1506

> | `includeList` is a list of metric names to specify which ones to generate. The names correspond to the names in Prometheus without the prefix. For example, `namespace_egress_packets_total` shows up as `netobserv_namespace_egress_packets_total` in Prometheus. Note that the more metrics you add, the bigger is the impact on Prometheus workload resources. Metrics enabled by default are: `namespace_flows_total`, `node_ingress_bytes_total`, `workload_ingress_bytes_total`, `namespace_drop_packets_total` (when `PacketDrop` feature is enabled), `namespace_rtt_seconds` (when `FlowRTT` feature is enabled), `namespace_dns_latency_seconds` (when `DNSTracking` feature is enabled). More information, with full list of available metrics: https://github.com/netobserv/network-observability-operator/blob/main/docs/Metrics.md

Author: After discussing this with the docs team, I've opened a PR against our contributing guidelines to make an exception for github links in regenerated API docs. So I think we can be good to go here with the GitHub link, and sorry for the worry.  No need to investigate avoiding github links in doc generation.
